File: minimum.java
// Time complexity : O(logn)
// Space complexity : O(1)
class Solution {
    public int findMin(int[] nums) {
        if(nums == null || nums.length == 0){
            return -1;
        }

        int low = 0;
        int high = nums.length-1;

        while (low <= high){
            if(nums[low] <= nums[high]){
                return nums[low];
            }
            int mid = low + (high - low)/2;

            // check if mid is less than its neighbours
            if((mid == 0 || nums[mid] < nums[mid-1]) && (mid == nums.length-1 || nums[mid]< nums[mid+1])){
                return nums[mid];
            }
            // check for left and right sorted
            if(nums[low] <= nums[mid]){
                //left sorted - reject 
                low = mid + 1;
            }
            else{
                high = mid-1;
            }
            
        }
        return -1;
    }
}

File: minimum.py
'''
Time Complexity: O(logn)
Space Complexity: O(1)
'''
class Solution:
    def findMin(self, nums: List[int]) -> int:
        low = 0
        high = len(nums)-1
        while(low<=high):
            mid = (low+high)//2
            if(nums[mid]>=nums[low] and nums[mid]<=nums[high]):
                return nums[low]
            elif(nums[mid]<nums[low]):
                high = mid
            elif(nums[mid]>nums[high]):
                if(low == mid and mid+1==high):
                    low = mid+1
                else:
                    low = mid
        return -1